core/src/main/java/novoda/lib/sqliteprovider/util/DatabaseStructure.java

```diff
@@ -0,0 +1,226 @@
+        Map<String, String> projection = new TreeMap<>();
+        for (String ft : foreignTables) {
+            projection.putAll(projectionFor(ft, columns(ft)));
+        }
+        return projection;
+    }
+
+    public List<String> foreignTables(String table) {
+        foreignTables.put(table, new ArrayList<String>(5));
+        final Cursor columnsCursor = queryTableColumnsFor(table);
+        List<String> allTables = tables();
+        parseForeignTablesFor(table, columnsCursor, allTables);
+        columnsCursor.close();
+        return Collections.unmodifiableList(foreignTables.get(table));
+    }
+
+    private void parseForeignTablesFor(String table, Cursor columnsCursor, List<String> allTables) {
+        while (columnsCursor.moveToNext()) {
+            String columnName = columnsCursor.getString(columnsCursor.getColumnIndexOrThrow(COLUMN_NAME));
+            addForeignTableIfExists(table, columnName, allTables);
+        }
+    }
+
+    private void addForeignTableIfExists(String forTable, String columnName, List<String> tables) {
+        if (!isForeignKey(columnName)) {
+            return;
+        }
+        String tableName = columnName.substring(0, columnName.lastIndexOf('_'));
+        if (tables.contains(tableName + "s")) {
+            foreignTables.get(forTable).add(tableName + "s");
+        } else if (tables.contains(tableName)) {
+            foreignTables.get(forTable).add(tableName);
+        }
+    }
+
+    private boolean isForeignKey(String columnName) {
+        return columnName.endsWith("_id");
+    }
+
+    public Map<String, SQLiteType> columns(String table) {
+        final Cursor columnsCursor = queryTableColumnsFor(table);
+        Map<String, SQLiteType> columns = parseColumns(columnsCursor);
+        columnsCursor.close();
+        return Collections.unmodifiableMap(columns);
+    }
+
+    private Map<String, SQLiteType> parseColumns(Cursor columnsCursor) {
+        Map<String, SQLiteType> columns = new HashMap<>(columnsCursor.getCount());
+        while (columnsCursor.moveToNext()) {
+            String name = columnsCursor.getString(columnsCursor.getColumnIndexOrThrow(COLUMN_NAME));
+            String type = columnsCursor.getString(columnsCursor.getColumnIndexOrThrow(COLUMN_TYPE));
+            columns.put(name, SQLiteType.fromName(type));
+        }
+        return columns;
+    }
+
+    private Cursor queryTableColumnsFor(String table) {
+        return database.rawQuery(String.format(PRAGMA_TABLE_INFO, table), null);
+    }
+
+    public List<Constraint> uniqueConstraints(String table) {
+        uniqueConstraints.put(table, new ArrayList<Constraint>());
+        lookForImplicitUniqueIndex(table);
+        final Cursor indexCursor = queryIndexListForTable(table);
+        parseUniqueConstraints(indexCursor, table);
+        indexCursor.close();
+        return Collections.unmodifiableList(uniqueConstraints.get(table));
+    }
+
+    private void lookForImplicitUniqueIndex(String table) {
```

Comment:
Possibly want to add a `For` suffix to the method name for consistency, here and for other 1-argument methods